lib/mongoid/shardable.rb

```diff
@@ -63,7 +63,7 @@ def shard_key_fields
     def shard_key_selector
       selector = {}
       shard_key_fields.each do |field|
-        selector[field.to_s] = new_record? ? send(field) : attribute_was(field)
+        selector[field.to_s] = new_record? || during_post_persist_callbacks ? send(field) : attribute_was(field)
```

Comment:
I don't think that this change introduces new bugs where the wrong value gets cached for `atomic_selector`. I think this because this change correctly returns the most-up-to-date value for each shard key at the time it is called. This means that if atomic_selector caches this value, the value is valid until the next time the shard key is updated.

However, you're probably correct that there are other instances where `atomic_selector` caching breaks things. I think this might be worth bringing up to Mongoid. I'm curious what real level of performance benefits Mongoid gets from this caching...